mlkem/poly.c

```diff
@@ -299,35 +299,52 @@ void poly_frommsg(poly *r, const uint8_t msg[MLKEM_INDCPA_MSGBYTES]) {
   POLY_BOUND_MSG(r, MLKEM_Q, "poly_frommsg output");
 }
 
-/*************************************************
- * Name:        poly_tomsg
- *
- * Description: Convert polynomial to 32-byte message
- *
- * Arguments:   - uint8_t *msg: pointer to output message
- *              - const poly *a: pointer to input polynomial
- *                  Coefficients must be unsigned canonical
- **************************************************/
 void poly_tomsg(uint8_t msg[MLKEM_INDCPA_MSGBYTES], const poly *a) {
   POLY_UBOUND(a, MLKEM_Q);
 
-  unsigned int i, j;
-  uint32_t t;
-
-  for (i = 0; i < MLKEM_N / 8; i++) {
-    msg[i] = 0;
-    for (j = 0; j < 8; j++) {
-      t = a->coeffs[8 * i + j];
-      // t += ((int16_t)t >> 15) & MLKEM_Q;
-      // t  = (((t << 1) + MLKEM_Q/2)/MLKEM_Q) & 1;
-      t <<= 1;
-      t += 1665;
-      t *= 80635;
-      t >>= 28;
-      t &= 1;
-      msg[i] |= t << j;
+  for (int i = 0; i < MLKEM_N / 8; i++)
+      // clang-format off
+    ASSIGNS(i, OBJECT_WHOLE(msg))
+    INVARIANT(i >= 0 && i <= MLKEM_N / 8)
+    DECREASES(MLKEM_N / 8 - i)
+    // clang-format on
+    {
+      msg[i] = 0;
+      for (int j = 0; j < 8; j++)
+          // clang-format off
+        ASSIGNS(j, OBJECT_WHOLE(msg))
+        INVARIANT(i >= 0 && i <= MLKEM_N / 8 && j >= 0 && j <= 8)
+        DECREASES(8 - j)
+        // clang-format on
+        {
+          uint32_t coeff = a->coeffs[8 * i + j];
+
+          // We now apply the "Compress1" function from FIPS-203 (eq 4.7)
+          // A simple expansion would be:
+          //   t += ((int16_t)t >> 15) & MLKEM_Q;
+          //   t  = (((t << 1) + MLKEM_Q/2)/MLKEM_Q) & 1;
+          // but we require constant-time evaluation, avoiding the division
+          // operator. We therefore use a Montgomery division, using magic
```

Comment:
That's not Montgomery, it's Barrett?

Reply:
The trick to doing division by a constant using only a multiplication and a shift owes to Montgomery - it's in PLDI'94. I see the above is doing a constant time division, not a reduction per se.

Reply:
We may be talking about different things. It may well be the case that the axiomitic definition in comments is an example of a Barrett Reduction. But where the code multiplies by 80635 and shifts right 28 places - that's a Granlund/Montgomery division.

Reply:
The "1665" should probably be "HALF_Q" though...